src/core_plugins/tagcloud/public/tag_cloud.js

```diff
@@ -238,69 +259,68 @@ class TagCloud extends EventEmitter {
     return mapSizeToFontSize;
   }
 
-  _makeJob() {
+  _makeNewJob() {
     return {
+      refreshLayout: true,
       words: this._words.map(toWordTag)
     };
   }
 
+  _makeJobPreservingLayout() {
+    return {
+      refreshLayout: false,
+      words: this._currentJob.words.map(tag => {
+        return {
+          x: tag.x,
+          y: tag.y,
+          rotate: tag.rotate,
+          size: tag.size,
+          text: tag.text
+        };
+      })
+    };
+  }
+
   _invalidate(keepLayout) {
 
     if (!this._words) {
       return;
     }
 
-    clearTimeout(this._handle);
-    this._handle = setTimeout(() => {
-      this._handle = null;
-      this._updateContainerSize();
-      if (keepLayout && this._currentJob && this._queue.length === 0) {
-        this._scheduleLayout({
-          words: this._currentJob.words.map(tag => {
-            return {
-              x: tag.x,
-              y: tag.y,
-              rotate: tag.rotate,
-              size: tag.size,
-              text: tag.text
-            };
-          })
-        });
-      } else {
-        this._updateLayout();
-      }
-    }, 0);//unhook from callstack. this avoids kicking off multiple layouts if multiple changes come in succession
+    this._updateContainerSize();
+    this._pendingJob = (!keepLayout || !this._currentJob || this._pendingJob) ? this._makeNewJob() : this._makeJobPreservingLayout();
+    this._processPendingJob();
   }
 
-  _scheduleLayout(job) {
-    this._queue.unshift(job);
-    this._processQueue();
-  }
 
-  _updateLayout() {
+  async _updateLayout(job) {
 
-    const job = this._makeJob();
     const mapSizeToFontSize = this._makeTextSizeMapper();
-
     const tagCloudLayoutGenerator = d3TagCloud();
     tagCloudLayoutGenerator.size(this._size);
     tagCloudLayoutGenerator.padding(this._padding);
     tagCloudLayoutGenerator.rotate(ORIENTATIONS[this._orientation]);
     tagCloudLayoutGenerator.font(this._fontFamily);
     tagCloudLayoutGenerator.fontStyle(this._fontStyle);
     tagCloudLayoutGenerator.fontWeight(this._fontWeight);
-    tagCloudLayoutGenerator.fontSize(tag => {
-      return mapSizeToFontSize(tag.value);
-    });
+    tagCloudLayoutGenerator.fontSize(tag => mapSizeToFontSize(tag.value));
     tagCloudLayoutGenerator.random(seed);
     tagCloudLayoutGenerator.spiral(this._spiral);
     tagCloudLayoutGenerator.words(job.words);
     tagCloudLayoutGenerator.text(getText);
     tagCloudLayoutGenerator.timeInterval(this._timeInterval);
-    tagCloudLayoutGenerator.on('end', () => this._scheduleLayout(job));
-    tagCloudLayoutGenerator.start();
+
+    this._layoutIsUpdating = true;
+    await new Promise((resolve) => {
```

Comment:
should this be returned? i see this being awaited on [here](https://github.com/elastic/kibana/pull/9445/files#diff-46088679e83da5567284a4c0f588ee15R132)